crates/sui-indexer/src/apis/governance_api.rs

```diff
@@ -205,104 +165,100 @@ impl<T: R2D2Connection + 'static> GovernanceReadApi<T> {
-    ) -> Result<Vec<ValidatorExchangeRates>, IndexerError> {
-        // Get validator rate tables
-        let mut tables = vec![];
+/// Cached exchange rates for validators for the given epoch, the cache size is 1, it will be cleared when the epoch changes.
+/// rates are in descending order by epoch.
+#[cached(
+    type = "SizedCache<EpochId, Vec<ValidatorExchangeRates>>",
+    create = "{ SizedCache::with_size(1) }",
+    convert = "{ system_state_summary.epoch }",
+    result = true
+)]
+pub async fn exchange_rates(
+    state: &GovernanceReadApi<impl R2D2Connection>,
+    system_state_summary: &SuiSystemStateSummary,
+) -> Result<Vec<ValidatorExchangeRates>, IndexerError> {
+    // Get validator rate tables
+    let mut tables = vec![];
+
+    for validator in &system_state_summary.active_validators {
+        tables.push((
+            validator.sui_address,
+            validator.staking_pool_id,
+            validator.exchange_rates_id,
+            validator.exchange_rates_size,
+            true,
+        ));
+    }
 
-        for validator in system_state_summary.active_validators {
-            tables.push((
-                validator.sui_address,
-                validator.staking_pool_id,
-                validator.exchange_rates_id,
-                validator.exchange_rates_size,
-                true,
-            ));
-        }
+    // Get inactive validator rate tables
+    for df in state
+        .inner
+        .get_dynamic_fields_in_blocking_task(
+            system_state_summary.inactive_pools_id,
+            None,
+            system_state_summary.inactive_pools_size as usize,
+        )
+        .await?
+    {
+        let pool_id: sui_types::id::ID = bcs::from_bytes(&df.bcs_name).map_err(|e| {
+            sui_types::error::SuiError::ObjectDeserializationError {
+                error: e.to_string(),
+            }
+        })?;
+        let inactive_pools_id = system_state_summary.inactive_pools_id;
+        let validator = state
+            .inner
+            .spawn_blocking(move |this| {
+                sui_types::sui_system_state::get_validator_from_table(
+                    &this,
+                    inactive_pools_id,
+                    &pool_id,
+                )
+            })
+            .await?;
+        tables.push((
+            validator.sui_address,
+            validator.staking_pool_id,
+            validator.exchange_rates_id,
+            validator.exchange_rates_size,
+            false,
+        ));
+    }
 
-        // Get inactive validator rate tables
-        for df in self
+    let mut exchange_rates = vec![];
+    // Get exchange rates for each validator
+    for (address, pool_id, exchange_rates_id, exchange_rates_size, active) in tables {
+        let mut rates = vec![];
+        for df in state
             .inner
-            .get_dynamic_fields_in_blocking_task(
-                system_state_summary.inactive_pools_id,
+            .get_dynamic_fields_raw_in_blocking_task(
```

Comment:
do we need to change the implementation here?

Reply:
This code got messed up over the time. The original implementation that is correct uses this code.
cc @bmwill

Reply:
Actually, the diff is messing things up. Everything is as it was before, except moving the function outside and adding a cache on it.